include/zephyr/drivers/pinctrl/pinctrl_kscan_it8xxx2.h

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2022 ITE Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_PINCTRL_PINCTRL_KSCAN_IT8XXX2_H_
+#define ZEPHYR_INCLUDE_DRIVERS_PINCTRL_PINCTRL_KSCAN_IT8XXX2_H_
+
+#include <pinctrl_kscan_soc.h>
+
+/**
+ * @brief Configure a set of pins.
+ *
+ * This function will configure the necessary hardware blocks to make the
+ * configuration immediately effective.
+ *
+ * @param pins List of pins to be configured.
+ * @param pin_cnt Number of pins.
+ *
+ * @retval 0 If succeeded
+ * @retval -errno Negative errno for other failures.
+ */
+int pinctrl_kscan_configure_pins(const struct pinctrl_kscan_soc_pin *pins,
+				 uint8_t pin_cnt);
+
```

Comment:
all this is messing with pinctrl/kscan in a custom/unconventional way. Please explain in detail why this is required, and why _standard_ mechanisms can't be used.

Reply:
Dear @gmarull,
 
According to the IT8XXX2 specification, the KSI[7:0] and KSO [15:0] control registers are defined in the chapter of Keyboard Scan (kscan) instead of the GPIO (pinctrl) for they use different hardware logic engines. Thus, the Keyboard Matrix pins have different register base from the GPIO’s, besides, the functionalities of the Keyboard Matrix pins are totally different from the ordinary GPIO pins. Meanwhile, the arrangements of the Keyboard Matrix control registers and of the GPIO control registers are totally different. Furthermore, the bit field definitions of Keyboard Matrix and GPIO control registers are also different so that the corresponding codes will also differ a lot. For example, if we’d like to set pull-up for any specific GPIO pin, we have to write the corresponding bit value in its corresponding control register for this specified pin. Assume that we’re trying to configure the GPIOD1 to be pulled-up, we have to set the bit 2 of the register 1629h. On the other hand, if we’d like to configure any pin in the KSO[15:0] to be pulled-up, we can set the bit 2 of the KSOCTRL (1D02h), but all the KSO[15:0] pins will be pulled-up simultaneously. In addition, as for the KSI[7:0] control registers, they don’t support open-drain. To put it differently, the KSO and KSI pins are totally different from GPIO pins because each GPIO pin control register supports multi-alternate mode selection, pull-up/pull-down, push-pull/open-drain, 1.8V/3.3V. Therefore, we hope that we can use two compatibles for Keyboard Matrix pins and  for GPIO pins respectively. It’s just like what we had done it in I2C driver before for the I2C0~2 and I2C3~5 are also using different hardware logic engines (#43357). What do you think?
 
Thanks for helps, guidance, and great patience.

Reply:
<p.s>
On our HW design, KSI[7:0]/KSO[15:0] belongs to keyboard scan module, KSO[17:16] belongs to GPIO module, so the `pinctrl_it8xxx2` driver doesn't suit for KSI[7:0]/KSO[15:0] (ex. pull-up/pull-down, push-pull/open-drain settings). If we use `pinctrl_it8xxx2` driver for KSI[7:0]/KSO[15:0] pins, then we would add more condition to configure keyboard scan or GPIO module settings, and the `pinctrl_it8xxx2` driver would be mess.

I tried that KSI[7:0]/KSO[15:0] use standard `pinctrl_configure_pins()`, but pinctrl-0 property and default state would get conflict between `pinctrl_it8xxx2` and `pinctrl_kscan_it8xxx2` driver. Pinctrl node doesn't like i2c that has multi-instance, so I add `pinctrl_kscan_configure_pins()` for `pinctrl_kscan_it8xxx2` driver.

Reply:
Hi @gmarull,
I create a PR #53391 that extended pinctrl driver for KSI/KSO pins.